src/fairchem/core/launchers/cluster/ray_cluster.py

```diff
@@ -0,0 +1,553 @@
+        except Exception as ex:
+            print(f"failed to load head info: {ex}. Maybe it's not ready yet?")
+            return None
+
+    def save_head_info(self, head_info: HeadInfo):
+        """
+        Saves the head node information to a JSON file.
+
+        Args:
+            head_info (HeadInfo): The head node information to save.
+        """
+        with self._head_json.open("w") as f:
+            json.dump(dataclasses.asdict(head_info), f)
+
+    def clean(self):
+        """Removes the rendezvous directory and all its contents."""
+        rmdir(self.rendezvous_dir)
+
+    def add_job(self, job: submitit.Job):
+        """
+        Adds a job to the jobs directory by creating a JSON file with the job's information.
+
+        Args:
+            job (submitit.Job): The job to add.
+        """
+        with (self.jobs_dir / f"{job.job_id}.json").open("w") as f:
+            json.dump(
+                {
+                    "job_id": job.job_id,
+                },
+                fp=f,
+            )
+
+    def list_job_ids(self) -> tp.List[str]:
+        """Lists all job IDs stored in the jobs directory."""
+        return [f.stem for f in self.jobs_dir.iterdir()]
+
+
+def _ray_head_script(cluster_state: RayClusterState, worker_wait_timeout_seconds: int):
+    """Start the head node of the Ray cluster on slurm."""
+    hostname = socket.gethostname()
+    head_env = os.environ.copy()
+    num_cpus = os.environ.get("SLURM_CPUS_ON_NODE", 1)
+    num_gpus = os.environ.get("SLURM_GPUS_ON_NODE", 0)
+    # using 0 as the port for the head will make ray search for an open port instead of
+    # always using the same one.
+    port = find_free_port()
+    head_env["RAY_ADDRESS"] = f"{hostname}:{port}"
+    head_env["RAY_gcs_server_request_timeout_seconds"] = str(
+        worker_wait_timeout_seconds
+    )
+    print(f"host {hostname}:{port}")
+    with tempfile.TemporaryDirectory(dir="/tmp") as temp_dir:
+        # ray workers have the same tempdir name (even on a different host)
+        # as the head. This is a problem when we use  /scratch/slurm_tmpdir/JOBID as
+        # the tempdir of the head job will not be accessible/visible from other workers if they
+        # are scheduled on the same host. We are forced to use a different tempdir than /scratch
+        # TODO ideally, we would still have a /scratch dir that everyone can share.
+        process = subprocess.Popen(
+            [
+                "ray",
+                "start",
+                "--head",
+                f"--port={port}",
+                f"--temp-dir={temp_dir}",
+                "--num-cpus",
+                f"{num_cpus}",
+                "--num-gpus",
+                f"{num_gpus}",
+                "--dashboard-host=0.0.0.0",
+            ],
+            env=head_env,
+            stdout=subprocess.PIPE,
+            text=True,
+        )
+        started = False
+        for line in stdout_lines(process):
```

Comment:
Is this function really adding anything? Cant we just do, 

for line in process.stdout:
?

Reply:
This isnt our code, its from fairray.